stacker/actions/graph.py

```diff
@@ -0,0 +1,68 @@
+import logging
+import sys
+import json
+
+from .base import BaseAction, plan
+
+
+logger = logging.getLogger(__name__)
+
+
+def each_step(graph):
+    """Returns an iterator that yields each step and it's direct
+    dependencies.
+    """
+
+    steps = graph.topological_sort()
+    steps.reverse()
+
+    for step in steps:
+        deps = graph.downstream(step.name)
+        yield (step, deps)
+
+
+def dot_format(out, graph, name="digraph"):
+    """Outputs the graph using the graphviz "dot" format."""
+
+    out.write("digraph %s {\n" % name)
+    for step, deps in each_step(graph):
+        for dep in deps:
+            out.write("  \"%s\" -> \"%s\";\n" % (step, dep))
+
+    out.write("}\n")
+
+
+def json_format(out, graph):
+    """Outputs the graph in a machine readable JSON format."""
+    steps = {}
+    for step, deps in each_step(graph):
+        steps[step.name] = {}
+        steps[step.name]["deps"] = [dep.name for dep in deps]
+
+    json.dump({"steps": steps}, out, indent=4)
+    out.write("\n")
+
+
+FORMATTERS = {
```

Comment:
I tend to throw all constants at the top of the file so they're immediately apparent - that said, just a personal preference.  Feel free to ignore :)

Reply:
Yeah, normally I would too, but that won't work in this case because were reference functions defined above:

```python
NameError: name 'dot_format' is not defined
```

Reply:
Ahh, right - ignore me :)